server/service/src/sensor/update.rs

```diff
@@ -54,6 +60,98 @@ pub fn update_sensor(
     Ok(sensor)
 }
 
+pub fn update_sensor_logs_for_breach(
+    connection: &StorageConnection,
+    breach_id: &String,
+) -> Result<Vec<TemperatureLog>, RepositoryError> {
+    let mut temperature_logs: Vec<TemperatureLog> = Vec::new();
+
+    let breach_result =
+        TemperatureBreachRowRepository::new(connection).find_one_by_id(breach_id)?;
+
+    if let Some(mut breach_record) = breach_result {
+        let is_cumulative_breach = (breach_record.r#type
+            == TemperatureBreachRowType::ColdCumulative)
+            | (breach_record.r#type == TemperatureBreachRowType::HotCumulative);
+        let logs = get_sensor_logs_for_breach(connection, breach_id)?; //sorted by date/time
+        let mut log_interval = 0;
+        let zero_time = NaiveTime::parse_from_str("00:00", "%H:%M").unwrap(); // hard-coded -> should always work!
+
+        if let Some(sensor) = SensorRowRepository::new(connection).find_one_by_id(&breach_record.sensor_id)? {
+            if let Some(interval) = sensor.log_interval {
+                log_interval = interval;
+            }
+        }
+        
+        if is_cumulative_breach {
+            // Update breach start/end from first/last logs if it has changed
+            if let Some(first_log) = logs.first() {
+                // If within log_interval seconds of the start of the day,
+                // use the start of the day as first log time
+                let mut first_breach_datetime = NaiveDateTime::new(first_log.temperature_log_row.datetime.date(), zero_time);
+                let first_log_datetime = first_breach_datetime + Duration::seconds(log_interval.into());
+
+                if first_log_datetime < first_log.temperature_log_row.datetime {
+                    first_breach_datetime = first_log.temperature_log_row.datetime;
+                }
+
+                // Use the first log time if it's either earlier than the calculated breach end
+                // or if the calculated breach start is less than the sensor log interval before the first log time
+                if (breach_record.start_datetime > first_breach_datetime) | (log_interval > 0) & (breach_record.start_datetime < first_breach_datetime - Duration::seconds(log_interval.into())) {
```

Comment:
I don't think it's an issue for Q-tags, but the raw data from Fridge tags doesn't actually have a breach end time (or a start time for that matter), just a duration and a trigger time, so the `temperature-sensor` crate tries to work it out. The problem is that this only works if it's a continuous cumulative breach (i.e. not split into more than one chunk during the day), so the calculated end time will _often_ be too early. As I discovered (see the testing section in #2594), the calculated end time can _sometimes_ also be too late. 

Hence it's more reliable to use the raw temperature log data, where it exists, to correct these miscalculations i.e. reset the breach start/end times when they aren't consistent with the first/last temperature logs 
I'm happy to create unit tests to validate that it all works.

That then leads to an additional adjustment to take account of the fact that temperature logs are only generated every X minutes, which means that you can easily end up with one cumulative breach having its last log at say 23:57 and the next cumulative breach having its first log at 00:02, when clearly it has all been continuous and the end/start time is actually midnight (there is an argument for combining these into a single cumulative breach which spans multiple days, but that's for the future).

I did consider doing these corrections inside the `temperature-sensor` crate, but it's much easier inside omSupply as all the data is in the DB.

Reply:
> and a trigger time, so the temperature-sensor crate tries to work it out. The problem is that this only works if it's a continuous cumulative breach

Ahh true, I think we should remove the end date time for cumulative breaches returned by `temperature-sensor` crate (just to avoid confusion).

> Hence it's more reliable to use the raw temperature log data, where it exists, to correct these miscalculations

Oki, in that case I agree, but can't we just use first and last log for the day for this ? (as simple as we can make it ?). And maybe split this into a new method just for that rather then a side effect of `update_sensor_logs_for_breach` ?

> That then leads to an additional adjustment 
> (there is an argument for combining these into a single cumulative breach which spans multiple days, but that's for the future).

Cumulative breaches never made sense to me, you need a threshold duration **and** a boundary, but the boundary is never defined, for fridge tags it's 24 hours from midnight to midnight (which makes little sense, since you can have a cumulative breach at night, but it won't be identified, and actually this is a likely scenario for power outage or for fridge unplugged, since no one is at the facility to fix it), in cold chain app the boundary is whatever 'dataset' you are looking at (the chart to and from date). What I am trying to get at here, is that it's a guessing game of 'what is the right' way to do cumulative breach, and we should just take fridge tag definition of it. (and not worry about special cases like continuous cumulative past midnight).

Saying all that, how does this sound:

For cumulative, use last log of the day that matches breach threshold, and set start and end date for this ? (start date should be know right ? even if it's not accurate, I think we should use it, to keep consistent with fridge tag data).

And of course associate every temperature log on the day that matches breach threshold with that cumulative breach (but of course do the consecutive first, and filter out not null breach ids for cumulative)

Reply:
> > and a trigger time, so the temperature-sensor crate tries to work it out. The problem is that this only works if it's a continuous cumulative breach
> 
> Ahh true, I think we should remove the end date time for cumulative breaches returned by `temperature-sensor` crate (just to avoid confusion).

I'm not quite sure that I follow?

> > Hence it's more reliable to use the raw temperature log data, where it exists, to correct these miscalculations
> 
> Oki, in that case I agree, but can't we just use first and last log for the day for this ? (as simple as we can make it ?). 

Even if we just use the temperature logs, then I think we should still do the midnight rounding (first two steps in https://github.com/msupply-foundation/open-msupply/pull/2629#discussion_r1488237477). 

Either way, the downside of using just the temperature logs is that you end up editing the start & end of _all_ the breaches, not just the few exceptions which have dodgy data, and you potentially discard some relevant data e.g. if you have an actual breach from 14:30-15:01 (say you have a threshold duration of 30 minutes), with breached logs at 14:32, 14:37 .. 14:57, then it appears that the breach was only 25 minutes (i.e. less than the threshold duration). Having said that, the stored breach duration will still be 31 minutes - that is independent of the start/end timestamps for cumulative breaches.

> And maybe split this into a new method just for that rather then a side effect of `update_sensor_logs_for_breach` ?

Fair enough - I can rename this method to something like `assign_sensor_logs_to_breach` and split the extra logic out into something like `update_breach_from_sensor_logs`

> > That then leads to an additional adjustment
> > (there is an argument for combining these into a single cumulative breach which spans multiple days, but that's for the future).
> 
> Cumulative breaches never made sense to me, you need a threshold duration **and** a boundary, but the boundary is never defined, for fridge tags it's 24 hours from midnight to midnight (which makes little sense, since you can have a cumulative breach at night, but it won't be identified, and actually this is a likely scenario for power outage or for fridge unplugged, since no one is at the facility to fix it), in cold chain app the boundary is whatever 'dataset' you are looking at (the chart to and from date). What I am trying to get at here, is that it's a guessing game of 'what is the right' way to do cumulative breach, and we should just take fridge tag definition of it. (and not worry about special cases like continuous cumulative past midnight).

I absolutely agree on the whole "how do you define a cumulative breach?" thing - for Berlinger tags, the inference from the data is that it's based on a 24-hour midnight to midnight period, but that's only for them and I can see that there's a _very_ good argument for working it out on demand for any given interval. However, I don't think the fridge breakdown overnight example works - unless the fridge fixes itself very soon after midnight, you're going to end up with two cumulative breaches, one for each day, and that's going to be noticed.

> Saying all that, how does this sound:
> 
> For cumulative, use last log of the day that matches breach threshold, and set start and end date for this ? (start date should be know right ? even if it's not accurate, I think we should use it, to keep consistent with fridge tag data).

Of course, the main point of fridge monitors is that they detect and flag breaches, and it really doesn't make much practical difference if we're a few minutes out either way. It just pains me to leave something "nearly right most of the time" when I can make it "right all of the time" with just a little bit of extra work!

> And of course associate every temperature log on the day that matches breach threshold with that cumulative breach (but of course do the consecutive first, and filter out not null breach ids for cumulative)

Absolutely, that's the core of this PR

Reply:
Fair enough - lets just leave it as the bare minimum "assume that the sensor data can be trusted" for now anyway. I'd still be keen to add in some (probably fairly extensive) comments in the code to explain the potential failings in the raw data, while I still have the context (sort-of) loaded - although maybe the best place for that is in the docs for the `temperature-sensor` crate?

Reply:
>  although maybe the best place for that is in the docs for the temperature-sensor crate?

Yeah I think that would be great, thank you !